src/lightning/fabric/plugins/precision/fp8_transformer_engine.py

```diff
@@ -0,0 +1,109 @@
+
+    .. warning::  This is an :ref:`experimental <versioning:Experimental API>` feature.
+
+    Args:
+        precision: The precision
+        recipe: Recipe for the DelayedScaling
+            `configuration <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transform
+            er_engine.common.recipe.DelayedScaling`__. In dict format or the dataclass format.
+    """
+
+    precision: Literal["8-mixed-transformer-engine"] = "8-mixed-transformer-engine"
+
+    def __init__(
+        self, recipe: Optional[Union[Dict[str, Any], "DelayedScaling"]] = None, replace_layers: bool = True
+    ) -> None:
+        if not torch.cuda.get_device_capability() >= (8, 1):
+            raise NotImplementedError("Your CUDA device does not support fp8 mixed precision.")
+        if not _TRANSFORMER_ENGINE_AVAILABLE:
+            raise ModuleNotFoundError(str(_TRANSFORMER_ENGINE_AVAILABLE))
+        from transformer_engine.common.recipe import DelayedScaling
+
+        if recipe is None:
+            recipe = DelayedScaling()
+        elif isinstance(recipe, dict):
+            if "fp8_format" in recipe:
+                from transformer_engine.common.recipe import Format
+
+                recipe["fp8_format"] = getattr(Format, recipe["fp8_format"])
+            recipe = DelayedScaling(**recipe)
+
+        self.recipe = recipe
+        self.replace_layers = replace_layers
+
+    @contextmanager
+    def forward_context(self) -> Generator[None, None, None]:
+        import transformer_engine.pytorch as te
+
+        with te.fp8_autocast(enabled=True, fp8_recipe=self.recipe):
+            yield
+
+    def convert_module(self, module: torch.nn.Module) -> torch.nn.Module:
+        # avoid converting if any is found. assume the user took care of it
+        if self.replace_layers and not any("transformer_engine" in m.__module__ for m in module.modules()):
+            _convert_layers(module)
+        return module
+        # TODO: should we un-convert on teardown?
+
+
+def _convert_layers(module: torch.nn.Module) -> None:
+    import transformer_engine.pytorch as te
+
+    for name, child in module.named_children():
+        if isinstance(child, torch.nn.Linear):
+            if child.in_features % 16 != 0 or child.out_features % 16 != 0:
+                # https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html#FP8-autocasting
+                rank_zero_warn(
+                    "Support for FP8 in the linear layers with `precision='8-mixed'` is currently limited to tensors"
+                    " with shapes where both dimensions are divisible by 16. You might want to add padding to your"
+                    " inputs so that you can fit this criteria."
+                )
+                continue
+            has_bias = child.bias is not None
+            replacement = te.Linear(child.in_features, child.out_features, bias=has_bias)
```

Comment:
Replacement instead of patching has a caveat in that isinstance checks will not pass, for example here: https://github.com/Lightning-AI/lit-parrot/blob/main/lit_parrot/model.py#L34

Reply:
does it make sense to subclass the few modules they implement and overwrite `__subclasscheck__` and `__instancecheck`  or would you say that isn't worth the effort?

Reply:
That would work, but might open the door to other issues.